src/lib/operators/join_hash/join_hash_steps.hpp

```diff
@@ -81,16 +84,16 @@ class PosHashTable {
   // of the offset does not limit the number of rows in the partition but the number of distinct values. If we end up
   // with a partition that has more values, the partitioning algorithm is at fault.
   using Offset = uint32_t;
-  using HashTable = ska::bytell_hash_map<HashedType, Offset>;
+  using HashTable = tsl::robin_pg_map<HashedType, Offset, std::hash<HashedType>, std::equal_to<HashedType>, PolymorphicAllocator<std::pair<HashedType, Offset>>>;
 
   // The small_vector holds the first n values in local storage and only resorts to heap storage after that. 1 is chosen
   // as n because in many cases, we join on primary key attributes where by definition we have only one match on the
   // smaller side.
-  using SmallPosList = boost::container::small_vector<RowID, 1>;
+  using SmallPosList = boost::container::small_vector<RowID, 1, PolymorphicAllocator<RowID>>; // TODO PMR raus
 
  public:
   explicit PosHashTable(const JoinHashBuildMode mode, const size_t max_size)
-      : _hash_table(), _pos_lists(max_size + 1), _mode(mode) {
+      : _buffer(std::make_shared<boost::container::pmr::monotonic_buffer_resource>()), _hash_table(&*_buffer), _pos_lists(mode == JoinHashBuildMode::AllPositions ? max_size + 1 : 0, SmallPosList{buffer.get()}), _mode(mode) {
```

Comment:
(3) Allocate memory for _pos_lists only if actually used
- Rename SinglePosition to Existence?

Reply:
https://github.com/hyrise/hyrise/compare/61bd7a4..3e9a33d

I only got this running in combination with the new hash map.

```
+----------++----------+---------+--------++----------+----------+--------+---------+
| Item     || Latency (ms/iter)  | Change || Throughput (iter/s) | Change | p-value |
|          ||      old |     new |        ||      old |      new |        |         |
+----------++----------+---------+--------++----------+----------+--------+---------+
| TPC-H 01 ||   1651.5 |  1685.3 |   +2%  ||     0.61 |     0.59 |   -2%  |  0.2545 |
| TPC-H 02 ||     11.8 |    11.0 |   -7%  ||    84.50 |    90.97 |   +8%  |  0.0000 |
| TPC-H 03 ||    178.4 |   177.2 |   -1%  ||     5.61 |     5.64 |   +1%  |  0.0010 |
| TPC-H 04 ||    152.0 |   139.0 |   -9%  ||     6.58 |     7.19 |   +9%  |  0.0000 |
| TPC-H 05 ||    470.9 |   460.5 |   -2%  ||     2.12 |     2.17 |   +2%  |  0.0000 |
| TPC-H 06 ||      7.9 |     8.0 |   +1%  ||   126.54 |   125.28 |   -1%  |  0.0001 |
| TPC-H 07 ||    126.2 |   118.7 |   -6%  ||     7.93 |     8.43 |   +6%  |  0.0000 |
| TPC-H 08 ||    156.3 |   147.9 |   -5%  ||     6.40 |     6.76 |   +6%  |  0.0000 |
| TPC-H 09 ||   1013.1 |  1014.1 |   +0%  ||     0.99 |     0.99 |   -0%  |  0.8100 |
| TPC-H 10 ||    511.9 |   505.2 |   -1%  ||     1.95 |     1.98 |   +1%  |  0.2031 |
| TPC-H 11 ||     16.5 |    16.4 |   -0%  ||    60.71 |    60.86 |   +0%  |  0.1266 |
| TPC-H 12 ||     90.0 |    84.9 |   -6%  ||    11.11 |    11.78 |   +6%  |  0.0000 |
| TPC-H 13 ||    705.9 |   769.9 |   +9%  ||     1.42 |     1.30 |   -8%  |  0.0000 |
| TPC-H 14 ||     38.8 |    31.2 |  -20%  ||    25.77 |    32.09 |  +25%  |  0.0000 |
| TPC-H 15 ||     17.6 |    17.5 |   -1%  ||    56.89 |    57.25 |   +1%  |  0.0000 |
| TPC-H 16 ||    178.3 |   173.1 |   -3%  ||     5.61 |     5.78 |   +3%  |  0.0000 |
| TPC-H 17 ||     54.4 |    54.2 |   -0%  ||    18.39 |    18.44 |   +0%  |  0.3120 |
| TPC-H 18 ||   1312.0 |  1322.9 |   +1%  ||     0.76 |     0.76 |   -1%  |  0.1524 |
| TPC-H 19 ||     75.3 |    76.1 |   +1%  ||    13.28 |    13.14 |   -1%  |  0.0000 |
| TPC-H 20 ||     31.7 |    29.6 |   -7%  ||    31.54 |    33.83 |   +7%  |  0.0000 |
| TPC-H 21 ||    822.2 |   866.2 |   +5%  ||     1.22 |     1.15 |   -5%  |  0.0000 |
| TPC-H 22 ||    126.3 |   123.1 |   -3%  ||     7.92 |     8.12 |   +3%  |  0.0000 |
+----------++----------+---------+--------++----------+----------+--------+---------+
| Sum      ||   7748.8 |  7831.8 |   +1%  ||          |          |        |         |
| Geomean  ||          |         |        ||          |          |   +3%  |         |
+----------++----------+---------+--------++----------+----------+--------+---------+
```

Reply:
So we see the Q14 improvement again. I can't see what's so special about that query.

What happened when you used the old hash map?

Reply:
no matching function for call to for _hash_table(&*_buffer) if _hash_table is a bytell_hash_map

Reply:
Did you change the allocator to PolymorphicAllocator though?

Reply:
i think so...
https://github.com/hyrise/hyrise/compare/61bd7a4..3e9a33d#diff-08c8844263a2390aea4c1de0ddfc6fe5R86-R87
is there another place where this needs to be changed?

Reply:
Nope, that should do it.

On nemea, the effect is not as pronounced. Looks like the binary alignment used in Q14 is somewhat on an edge on your machine. I'd still say we should add this iff we add the tsl map.

```diff
 +----------++----------+---------+--------++----------+----------+--------+---------+
 | Item     || Latency (ms/iter)  | Change || Throughput (iter/s) | Change | p-value |
 |          ||      old |     new |        ||      old |      new |        |         |
 +----------++----------+---------+--------++----------+----------+--------+---------+
 | TPC-H 01 ||    821.3 |   817.1 |   -1%  ||     1.22 |     1.22 |   +1%  |  0.4644 |
 | TPC-H 02 ||      4.6 |     4.5 |   -1%  ||   218.39 |   219.92 |   +1%  |  0.3613 |
 | TPC-H 03 ||     81.4 |    81.0 |   -1%  ||    12.28 |    12.35 |   +1%  |  0.0001 |
 | TPC-H 04 ||     65.9 |    65.5 |   -1%  ||    15.18 |    15.26 |   +1%  |  0.0000 |
 | TPC-H 05 ||    223.3 |   223.5 |   +0%  ||     4.48 |     4.47 |   -0%  |  0.8261 |
 | TPC-H 06 ||      3.2 |     3.2 |   -1%  ||   310.77 |   312.72 |   +1%  |  0.0000 |
 | TPC-H 07 ||     57.0 |    57.1 |   +0%  ||    17.54 |    17.52 |   -0%  |  0.8625 |
 | TPC-H 08 ||     71.7 |    72.0 |   +0%  ||    13.95 |    13.89 |   -0%  |  0.3386 |
 | TPC-H 09 ||    508.6 |   505.6 |   -1%  ||     1.97 |     1.98 |   +1%  |  0.0862 |
 | TPC-H 10 ||    249.0 |   248.5 |   -0%  ||     4.02 |     4.02 |   +0%  |  0.8167 |
 | TPC-H 11 ||      8.3 |     8.2 |   -1%  ||   120.62 |   121.60 |   +1%  |  0.0000 |
 | TPC-H 12 ||     42.4 |    42.2 |   -0%  ||    23.60 |    23.68 |   +0%  |  0.0096 |
 | TPC-H 13 ||    354.9 |   350.5 |   -1%  ||     2.82 |     2.85 |   +1%  |  0.0000 |
 | TPC-H 14 ||     21.2 |    21.1 |   -1%  ||    47.06 |    47.46 |   +1%  |  0.0000 |
 | TPC-H 15 ||      7.7 |     7.7 |   -1%  ||   129.04 |   130.21 |   +1%  |  0.0000 |
 | TPC-H 16 ||     91.2 |    90.6 |   -1%  ||    10.96 |    11.03 |   +1%  |  0.0091 |
 | TPC-H 17 ||     16.4 |    16.3 |   -1%  ||    60.87 |    61.49 |   +1%  |  0.0000 |
+| TPC-H 18 ||    731.0 |   669.1 |   -8%  ||     1.37 |     1.49 |   +9%  |  0.0000 |
 | TPC-H 19 ||     27.8 |    27.7 |   -0%  ||    35.95 |    36.10 |   +0%  |  0.0000 |
 | TPC-H 20 ||     15.7 |    15.6 |   -0%  ||    63.80 |    64.01 |   +0%  |  0.1913 |
 | TPC-H 21 ||    414.8 |   410.2 |   -1%  ||     2.41 |     2.44 |   +1%  |  0.0218 |
 | TPC-H 22 ||     50.0 |    49.4 |   -1%  ||    20.01 |    20.24 |   +1%  |  0.0000 |
 +----------++----------+---------+--------++----------+----------+--------+---------+
 | Sum      ||   3867.4 |  3786.5 |   -2%  ||          |          |        |         |
 | Geomean  ||          |         |        ||          |          |   +1%  |         |
 +----------++----------+---------+--------++----------+----------+--------+---------+
```

Reply:
```diff
 +----------++----------+---------+--------++----------+----------+--------+---------+
 | Item     || Latency (ms/iter)  | Change || Throughput (iter/s) | Change | p-value |
 |          ||      old |     new |        ||      old |      new |        |         |
 +----------++----------+---------+--------++----------+----------+--------+---------+
 | TPC-H 01 ||    810.9 |   821.4 |   +1%  ||     1.23 |     1.22 |   -1%  |  0.0381 |
+| TPC-H 02 ||      5.2 |     4.9 |   -6%  ||   193.44 |   205.26 |   +6%  |  0.0000 |
 | TPC-H 03 ||     83.8 |    83.2 |   -1%  ||    11.93 |    12.02 |   +1%  |  0.0000 |
+| TPC-H 04 ||     68.4 |    62.6 |   -9%  ||    14.62 |    15.98 |   +9%  |  0.0000 |
+| TPC-H 05 ||    232.3 |   210.6 |   -9%  ||     4.30 |     4.75 |  +10%  |  0.0000 |
+| TPC-H 06 ||      3.5 |     3.3 |   -6%  ||   282.34 |   300.05 |   +6%  |  0.0000 |
 | TPC-H 07 ||     58.5 |    56.3 |   -4%  ||    17.08 |    17.77 |   +4%  |  0.0000 |
 | TPC-H 08 ||     76.5 |    74.1 |   -3%  ||    13.07 |    13.49 |   +3%  |  0.0000 |
 | TPC-H 09 ||    517.0 |   523.2 |   +1%  ||     1.93 |     1.91 |   -1%  |  0.0012 |
+| TPC-H 10 ||    262.4 |   248.6 |   -5%  ||     3.81 |     4.02 |   +6%  |  0.0000 |
 | TPC-H 11 ||      8.9 |     9.0 |   +1%  ||   112.24 |   111.00 |   -1%  |  0.0000 |
 | TPC-H 12 ||     43.6 |    41.9 |   -4%  ||    22.96 |    23.86 |   +4%  |  0.0000 |
-| TPC-H 13 ||    351.5 |   374.2 |   +6%  ||     2.85 |     2.67 |   -6%  |  0.0000 |
+| TPC-H 14 ||     21.1 |    17.2 |  -18%  ||    47.39 |    58.12 |  +23%  |  0.0000 |
 | TPC-H 15 ||      7.7 |     7.7 |   +0%  ||   130.28 |   129.74 |   -0%  |  0.0000 |
 | TPC-H 16 ||     93.8 |    91.2 |   -3%  ||    10.66 |    10.97 |   +3%  |  0.0000 |
 | TPC-H 17 ||     19.5 |    19.5 |   +0%  ||    51.38 |    51.17 |   -0%  |  0.0396 |
 | TPC-H 18 ||    697.6 |   682.7 |   -2%  ||     1.43 |     1.46 |   +2%  |  0.0371 |
 | TPC-H 19 ||     31.4 |    31.0 |   -1%  ||    31.89 |    32.27 |   +1%  |  0.0000 |
+| TPC-H 20 ||     16.5 |    15.4 |   -6%  ||    60.63 |    64.80 |   +7%  |  0.0000 |
-| TPC-H 21 ||    420.9 |   443.0 |   +5%  ||     2.38 |     2.26 |   -5%  |  0.0000 |
 | TPC-H 22 ||     50.5 |    49.2 |   -3%  ||    19.79 |    20.31 |   +3%  |  0.0000 |
 +----------++----------+---------+--------++----------+----------+--------+---------+
 | Sum      ||   3881.3 |  3870.2 |   -0%  ||          |          |        |         |
 | Geomean  ||          |         |        ||          |          |   +3%  |         |
 +----------++----------+---------+--------++----------+----------+--------+---------+
```